File: src/main/java/de/pflugradts/passbird/application/boot/main/PassbirdApplication.java
package de.pflugradts.passbird.application.boot.main;

import com.google.inject.Inject;
import com.google.inject.Singleton;
import de.pflugradts.passbird.application.UserInterfaceAdapterPort;
import de.pflugradts.passbird.application.boot.Bootable;
import de.pflugradts.passbird.application.commandhandling.InputHandler;
import de.pflugradts.passbird.application.failurehandling.FailureCollector;
import de.pflugradts.passbird.domain.model.transfer.Bytes;
import de.pflugradts.passbird.domain.model.transfer.Input;
import de.pflugradts.passbird.domain.model.transfer.Output;
import de.pflugradts.passbird.domain.service.NamespaceService;

import static de.pflugradts.passbird.domain.model.namespace.Namespace.DEFAULT;

@Singleton
public class PassbirdApplication implements Bootable {

    public static final char INTERRUPT = 0x03;

    @Inject
    private FailureCollector failureCollector;
    @Inject
    private UserInterfaceAdapterPort userInterfaceAdapterPort;
    @Inject
    private NamespaceService namespaceService;
    @Inject
    private InputHandler inputHandler;

    @Override
    public void boot() {
        userInterfaceAdapterPort.sendLineBreak();
        Input input;
        while (!isSigTerm(input = receiveInput())) {
            inputHandler.handleInput(input);
        }
    }

    private Input receiveInput() {
        return userInterfaceAdapterPort
                .receive(Output.Companion.outputOf(Bytes.bytesOf(namespacePrefix() + "Enter command: ")));
    }

    private String namespacePrefix() {
        final var currentNamespace = namespaceService.getCurrentNamespace();
        return currentNamespace.equals(DEFAULT) ? "" : "[" + currentNamespace.getBytes().asString() + "] ";
    }

    private boolean isSigTerm(final Input input) {
        return input.getData().isEmpty()
            && !input.getCommand().isEmpty()
            && input.getCommand().getFirstByte() == INTERRUPT;
    }

}

File: src/main/kotlin/de/pflugradts/passbird/application/boot/main/PassbirdApplication.kt
package de.pflugradts.passbird.application.boot.main

import de.pflugradts.passbird.application.RunContext
import de.pflugradts.passbird.application.UserInterfaceAdapterPort
import de.pflugradts.passbird.application.boot.Bootable
import de.pflugradts.passbird.application.commandhandling.InputHandler
import de.pflugradts.passbird.application.process.Initializer
import de.pflugradts.passbird.application.process.inactivity.InactivityHandler
import de.pflugradts.passbird.domain.model.nest.Nest
import de.pflugradts.passbird.domain.model.shell.Shell.Companion.shellOf
import de.pflugradts.passbird.domain.model.transfer.Input
import de.pflugradts.passbird.domain.model.transfer.Output.Companion.outputOf
import de.pflugradts.passbird.domain.model.transfer.OutputFormatting.NEST
import de.pflugradts.passbird.domain.service.nest.NestService
import jakarta.inject.Inject
import jakarta.inject.Singleton

const val INTERRUPT = 0x03.toChar()

@Singleton
class PassbirdApplication @Inject constructor(
    private val inactivityHandler: InactivityHandler,
    private val initializers: Set<Initializer>,
    private val inputHandler: InputHandler,
    private val nestService: NestService,
    private val runContext: RunContext,
    private val userInterfaceAdapterPort: UserInterfaceAdapterPort,
) : Bootable {

    override fun boot() {
        userInterfaceAdapterPort.sendLineBreak()
        nestService.moveToNestAt(runContext.initialSlot)
        initializers.forEach { it.run() }
        var input: Input
        while (!isSigTerm(receiveInput().also { input = it })) {
            inactivityHandler.registerInteraction()
            inputHandler.handleInput(input)
        }
    }

    private fun receiveInput() = userInterfaceAdapterPort.receive(
        outputOf(shellOf(nestPrefix()), NEST),
        outputOf(shellOf("Enter command: ")),
    )

    private fun nestPrefix() = nestService.currentNest().let {
        if (it == Nest.DEFAULT) "" else "[${it.viewNestId().asString()}] "
    }

    private fun isSigTerm(input: Input) = input.data.isEmpty && !input.command.isEmpty && input.command.firstByte == INTERRUPT.code.toByte()
}
